packages/presets/client/src/fragment-masking-plugin.ts

```diff
@@ -13,10 +24,43 @@ export type FragmentType<TDocumentType extends DocumentTypeDecoration<any, any>>
   : never;`;
 
 const makeFragmentDataHelper = `
+type UnionToIntersectGroupByTypeName<U, V = U> = [V] extends [
+  { __typename?: infer TypeName }
+]
+  ? TypeName extends any
+    ? UnionToIntersection<U extends { __typename?: TypeName } ? U : never>
+    : never
+  : never;
```

Comment:
We want to make the union of fragment type into an intersection type of them if and only if they have same `__typename` value.